source/extensions/transport_sockets/tls/ocsp/ocsp.h

```diff
@@ -0,0 +1,324 @@
+
+  const CertId cert_id_;
+  const CertStatus status_;
+  const Envoy::SystemTime this_update_;
+  const absl::optional<Envoy::SystemTime> next_update_;
+};
+
+/**
+ * Reflection of the ASN.1 ResponseData structure.
+ * Contains an OCSP response for each certificate in a given request
+ * as well as the time at which the response was produced.
+ */
+struct ResponseData {
+  ResponseData(Envoy::SystemTime produced_at, std::vector<SingleResponse> single_responses);
+
+  const Envoy::SystemTime produced_at_;
+  const std::vector<SingleResponse> single_responses_;
+};
+
+/**
+ * An abstract type for OCSP response formats. Which variant of |Response| is
+ * used in an |OcspResponse| is indicated by the structure's OID.
+ *
+ * We currently enforce that OCSP responses must be for a single certificate
+ * only. The methods on this class extract the relevant information for the
+ * single certificate contained in the response.
+ */
+class Response {
+public:
+  virtual ~Response() = default;
+
+  /**
+   * @return The number of certs reported on by this response.
+   */
+  virtual size_t getNumCerts() PURE;
+
+  /**
+   * @return The revocation status of the certificate.
+   */
+  virtual CertStatus getCertRevocationStatus() PURE;
+
+  /**
+   * @return The serial number of the certificate.
+   */
+  virtual const std::string& getCertSerialNumber() PURE;
+
+  /**
+   * @return The beginning of the validity window for this response.
+   */
+  virtual const Envoy::SystemTime& getThisUpdate() PURE;
+
+  /**
+   * The time at which this response is considered to expire. If
+   * |nullopt|, then there is assumed to always be more up-to-date
+   * information available and the response is always considered expired.
+   *
+   * @return The end of the validity window for this response.
+   */
+  virtual const absl::optional<Envoy::SystemTime>& getNextUpdate() PURE;
+};
+
+using ResponsePtr = std::unique_ptr<Response>;
+
+/**
+ * Reflection of the ASN.1 BasicOcspResponse structure.
+ * Contains the full data of an OCSP response and a signature/signature
+ * algorithm to verify the OCSP responder.
+ *
+ * BasicOcspResponse is the only supported Response type in RFC 6960.
+ */
+class BasicOcspResponse : public Response {
+public:
+  BasicOcspResponse(ResponseData data, std::string signature_alg, std::vector<uint8_t> signature);
+
+  // Response
+  size_t getNumCerts() override { return data_.single_responses_.size(); }
+  CertStatus getCertRevocationStatus() override { return data_.single_responses_[0].status_; }
+  const std::string& getCertSerialNumber() override {
+    return data_.single_responses_[0].cert_id_.serial_number_;
+  }
+  const Envoy::SystemTime& getThisUpdate() override {
+    return data_.single_responses_[0].this_update_;
+  }
+  const absl::optional<Envoy::SystemTime>& getNextUpdate() override {
+    return data_.single_responses_[0].next_update_;
+  }
+
+  const static std::string OID;
```

Comment:
static class initialization in c++ is a wreck. Either:

* make into constexpr string_view (best option if it works)
* make a static function to access in wrap in `CONSTRUCT_ON_FIRST_USE`